docs(manifest): Improvements to `name`, `short_name`, `start_url` (#35847)
mdn/content: files/en-us/web/manifest/start_url/index.md

{{QuickLinksWithSubpages("/en-US/docs/Web/Manifest")}}

> The `start_url` manifest member is used to specify the URL that should be opened when a user launches your web application, such as when tapping the application's icon on their device's home screen or in an application list.

Reviewer: I wonder if we should forward reference to "Browsers have flexibility in how they handle `start_url`."

It is pretty significant that your start URL might not be used. 

Maybe (?)

```suggestion
The `start_url` manifest member is a hint to the browser for the URL that should be opened when a user launches your web application, such as when tapping the application's icon on their device's home screen or in an application list.
```

Author: It was a note in the previous version. I've now added a note and linked to the subsection (added one now) in "Description"

> The `start_url` manifest member is used to specify the URL that should be opened when a user launches your web application, such as when tapping the application's icon on their device's home screen or in an application list.

> [!NOTE]
> [Browsers have flexibility](#browser_flexibility) in how they the handle `start_url` and may not always use the specified value.

Author: Had to update the note to avoid a link right at the beginning. `[!NOTE]` and `[` of the link right after seemed to be conflicting.

> [!NOTE]
> The `start_url` is a hint for browsers. [Browsers have flexibility](#browser_flexibility) in how they the handle `start_url` and may not always use the specified value.

> [!NOTE]
> The `start_url` is a hint for browsers. [Browsers have flexibility](#description) in how they the handle `start_url` and may not always use the specified value.

## Syntax

```json-nolint
/* Absolute URLs */
"start_url": "https://example.com/myapp"
"start_url": "https://myapp.com/home"

/* Relative URLs */
"start_url": "/"
"start_url": "../index.html"
```

### Values

- `start_url`

  - : A string that represents the starting URL of a web app.
    The URL can be absolute or relative.
    If the value is relative, it is resolved against the manifest URL.

    The `start_url` must be within the [scope](/en-US/docs/Web/Manifest/scope) of the web app and must be same-origin with the manifest URL.

    If `start_url` is unspecified or the value is invalid (i.e., not a string, not a valid URL, or not same-origin with the manifest URL), the URL of the page that links to the manifest is used.

## Description

The `start_url` allows you to recommend an appropriate common entry point for all users.

> Note however that browsers are not required to use the specified URL.
> They might ignore the specified value or provide users with alternative choices.
> The might also allow users to modify the URL at a later time, such as when creating a bookmark for the web app.
> Keep this in mind when designing your app .

Author: - Fixing few typos plus keeping some text similar to the spec.
- From the spec: "The [start_url](https://w3c.github.io/manifest/#dfn-start_url) member is purely advisory, and a user agent MAY [ignore](https://w3c.github.io/manifest/#dfn-ignore) it or provide the end-user the choice not to make use of it. A user agent MAY also allow the end-user to modify the URL when, for instance, a bookmark for the web application is being created or any time thereafter."
```suggestion
Note, however, that browsers are not required to use the specified URL.
They may ignore the specified value or provide users with a choice not to use it.
They may also allow users to modify the URL when creating a bookmark for the web app or at a later time.
Keep this in mind when designing your app to allow for variations in `start_url`.
```

{{QuickLinksWithSubpages("/en-US/docs/Web/Manifest")}}

The `start_url` manifest member is used to specify the URL that should be opened when a user launches your web application, such as when tapping the application's icon on their device's home screen or in an application list.

> [!NOTE]

> > The `start_url` is a hint for browsers. [Browsers have flexibility](#browser_flexibility) in how they the handle `start_url` and may not always use the specified value.

Reviewer: ```suggestion
> The `start_url` is a hint for browsers. [Browsers have flexibility](#description) in how they the handle `start_url` and may not always use the specified value.
```

## Syntax

```json-nolint
/* Absolute URLs */
"start_url": "https://example.com/myapp"
"start_url": "https://myapp.com/home"

/* Relative URLs */
"start_url": "/"
"start_url": "../index.html"
```

### Values

- `start_url`

  - : A string that represents the starting URL of a web app.
    The URL can be absolute or relative.
    If the value is relative, it is resolved against the manifest URL.

    The `start_url` must be within the [scope](/en-US/docs/Web/Manifest/scope) of the web app and must be same-origin with the manifest URL.

    If `start_url` is unspecified or the value is invalid (i.e., not a string, not a valid URL, or not same-origin with the manifest URL), the URL of the page that links to the manifest is used.

> ## Description
> 
> ### Best practices
> 
> To ensure the same entry point to your app for all users, specify a `start_url`.
> This URL should navigate users to an important page of your app, such as a dashboard.
> Consider features that users would want to access immediately after launching the app.
> If your app's main page is at the root of your site, you can set the `start_url` to `/`.
> You can also specify a deep link (e.g., `https://myapp.com/product/whatsnew`) to direct users to specific content within your app.
> Avoid specifying a generic starting page.
> 
> For security reasons, the `start_url` must be same-origin with the manifest URL.
> If a non-same-origin `start_url` is specified, browsers will fallback to using the page that links to the manifest as the default starting page.
> 
> ### Browser flexibility
> 
> Browsers may ignore the specified `start_url` value or provide users with a choice not to use it.
> Browsers may allow users to modify the URL, such as when creating a bookmark for the web app or at any later time.
> As a result, the `start_url` you specify may not be the one ultimately used.
> It is important for you to keep this in mind while designing your app to allow for variations in `start_url`.

Reviewer: I don't love the title "browser flexibility" and I think this info could be near the top. I'm going to do it like this and then merge:

```suggestion
## Description

The `start_url` allows you to recommend an appropriate common entry point for all users.

Note however that browsers are not required to use the specified URL.
They might ignore the specified value or provide users with alternative choices.
The might also allow users to modify the URL at a later time, such as when creating a bookmark for the web app.
Keep this in mind when designing your app .

### Best practices

This URL should navigate users to an important page of your app, such as a dashboard.
Consider features that users would want to access immediately after launching the app.
If your app's main page is at the root of your site, you can set the `start_url` to `/`.
You can also specify a deep link (e.g., `https://myapp.com/product/whatsnew`) to direct users to specific content within your app.
Avoid specifying a generic starting page.

For security reasons, the `start_url` must be same-origin with the manifest URL.
If a non-same-origin `start_url` is specified, browsers will fallback to using the page that links to the manifest as the default starting page.
```

{{QuickLinksWithSubpages("/en-US/docs/Web/Manifest")}}

The `start_url` manifest member is used to specify the URL that should be loaded when a user launches your web application, such as when tapping the app's icon on the home screen or in a device's application menu.

## Syntax

```json-nolint
/* Absolute URLs */
"start_url": "https://example.com/myapp"
"start_url": "https://myapp.com/home"

/* Relative URLs */
"start_url": "/"
"start_url": "../index.html"
```

### Values

> - `start_url`

Reviewer: ```suggestion
- `start_url` {{optional_inline}}
```

Author: All manifest members are optional: https://developer.mozilla.org/en-US/docs/Web/Manifest#members.
Do you think adding the optional tag to all would make this more explicit/clear?

Reviewer: Not certain. It is odd to add them in cases like this where the value is optional, in particular when the top level keys are optional. On the other hand, there are cases where the key is optional, but it is an object that has mandatory fields if the option is defined.

If it were me I would NOT add them for these cases.

  - : A string that represents the starting URL of a web app.
    The value must be same-origin with the page that links to the manifest file.
    If the value is relative, it is resolved against the manifest file's URL.

>     If `start_url` is unspecified or the value is invalid (i.e., not a string, not a valid URL, or not same-origin with the linking page), the URL of the page that links to the manifest file is used as a fallback.

Reviewer: "the URL of the page that links to the manifest file is used as a fallback." - only one page can link to the manifest file?

Reviewer: If so, is there a link we can provide to this info.

Author: I think I might have got the fallback behavior wrong. I am trying to get some clarification in https://github.com/w3c/manifest/issues/1147.

Author: Until the confusion about "document URL" as used in spec is cleared, I'm reverting to the previous version but with a slight change. Instead of "same-origin with page that links to manifest file", I now have "same-origin with the manifest URL". This is more plausible. 

> only one page can link to the manifest file?

Multiple pages in the app can link to the manifest file. It seems that "URL of the page that links to the manifest file" will eventually be the page from which the app is installed but I guess we'll need to wait for a response to my question in https://github.com/w3c/manifest/issues/1147.

fyi, the info was originally added via https://github.com/mdn/content/pull/34638/files but there is no explanation about the source.

Author: Updated the Description and Examples accordingly

Reviewer: That makes sense to me.

## Description

The `start_url` must be same-origin with the page that links to the manifest file for security reasons. If the URL is missing or invalid, the linking page is used as a fallback to ensure there's always a valid starting point for the app.

Browsers have flexibility in how they handle `start_url`. They may:

- Ignore the specified value.
- Provide users with a choice not to use it.
- Allow users to inspect and modify the URL, such as when creating a bookmark for the web app or any time later. Therefore, when developing your app, keep in mind that the `start_url` may be altered. This flexibility provided by browsers can help protect user privacy. See [Privacy considerations](#privacy_considerations) for more details.

> [!NOTE]

> > For Chromium-based browsers, a `start_url` is required in the manifest for the web app to be installable.

Reviewer: We have these notes everywhere. IMO this is compatibility information that should be moved to the BCD and removed from these docs (once that happens).

Author: I've opened https://github.com/mdn/browser-compat-data/pull/24700

Reviewer: Thanks, so would we remove this text now or wait until the BCD merges?

  Browsers may offer protection against this type of fingerprinting.
  For example, when a user clears data from an origin, the browser may prompt users to uninstall apps that are within the scope of that origin.
  This would remove any potential fingerprint from the app's `start_url`.

- **Launch tracking**:

  You can potentially add a `start_url` to indicate that the app was launched from outside the browser (e.g., `"start_url": "index.html?launcher=homescreen"`). While this can be useful for analytics and other customizations, be cautious about potential privacy implications. This information could be used as part of a user's digital fingerprint.

## Examples

### Specifying an absolute starting URL

Let's say the manifest file for your web app is linked from `https://example.com/index.html`.
You want a custom starting point when users launch the installed app from their devices, which is different from the default entry page for your web app.
You can specify this starting URL in your manifest file like so:

```json
"start_url": "https://example.com/dashboard.html"
```

The `start_url` value above is valid because it is same-origin with `https://example.com/index.html`.

> In contrast, the following `start_url` will be considered invalid:

Reviewer: ```suggestion
In contrast, the following `start_url` is invalid:
```

```json example-bad
"start_url": "https://other-domain.com/dashboard.html"
```

> In this case, the app launch will fallback to using `https://example.com/index.html`.

Reviewer: ```suggestion
```
